### DISCUSSION TITLE
JWT token security concerns

### DISCUSSION THREAD

**Discussion by sebpalluel** [nhost/nhost#738] [General]
From what i see [here](https://github.com/nhost/nhost/blob/a6cfdb67d0042d50af63f6e755d119a7734c8d1f/packages/core/src/storage.ts), the JWT token is stored in the localStorage.

According to [this article](https://hasura.io/blog/best-practices-of-using-jwt-with-graphql) it should be stored in the sessionStorage instead to reduce the risk of an XSS attack by avoiding to store the token on the client persistently. As described on the `Persisting sessions` section of the article, it would break the functionality of persisting a session when relaunching the browser.

There is also a mention about a fingerprint added to the token to mitigate the risk of token side-jacking. Is it implemented in `hasura-auth` or is there any plan to support it ?

[The sample project from hasura that implement thoses recommendations can be found here](https://github.com/hasura/jwt-guide). You can find the code generating the fingerprint cookie for the user [here](https://github.com/hasura/jwt-guide/blob/c66de6d5fb41ff8bba2bc4973df4cc6625a01e82/pages/api/actions/login.ts) and the fingerprint check for the refreshToken [here](https://github.com/hasura/jwt-guide/blob/c66de6d5fb41ff8bba2bc4973df4cc6625a01e82/pages/api/actions/refresh-jwt.ts)

**Comment by elitan** [comment]
The risk of XSS is very small in modern js frameworks like React, Vue, Svelte, etc. Also, if you have XSS vulnerabilities in your app I don't see how `sessionStorage` would save you?

If you're concerned about where to persistently store you can always overload our setters and getters and store the refresh token as you prefer:

https://github.com/nhost/nhost/blob/main/packages/nhost-js/src/core/nhost-client.ts#L37-L40

Maybe @plmercereau could provide an example of how to do that?

**Comment by plmercereau** [comment]
Hello Sébastien,

First of all, a clarification: the Nhost SDK does **not** store the JWT but the refresh token in localStorage. We don't store JWTs anywhere else than in memory.
When a user loads a page, it looks for a _refresh_ token in localStorage. This refresh token is sent to the backend (hasura-auth), that updates its expiration date, generates an _access_ token (JWT) and returns it to the client.
You cannot use this refresh token to authenticate any request sent to Hasura. If it has somehow been exposed, it can be invalidated in the backend so it won't be possible to generate any access token (JWT) from it.

We could implement a refresh token fingerprint mechanism, but that being said, we've made the choice to allow users to be able to re-open their session after closing the browser, and there is no other means to achieve that than persisting data in localStorage or in a cookie.

Anyhow, you can store the refresh token in sessionStorage:

```js
const nhost = new NhostClient({
  backendUrl: 'myapp.nhost.app',
  clientStorageType: 'custom',
  clientStorage: {
    getItem: key => sessionStorage.getItem(key) ?? null,
    setItem: sessionStorage.setItem,
    removeItem: sessionStorage.removeItem,
  },
})
```

**Comment by sebpalluel** [comment]
Thank you very much for both of your answers. I get that sessionStorage is not a major concern and more or less a choice to favor a bit security despite of UX. Thanks for the tip by the way.

I haven't seen a fingerprint mechanism in many implementation I've checked of JWT but for what I've read, mainly Redis and OWASP.org it's really a huge plus to mitigate some vector of attacks.

For what I understand [on the code from hasura here](https://github.com/hasura/jwt-guide/blob/master/ADDITIONS.md#token-sidejacking) it doesn't impact the UX :

1. The fingerprint string is stored in an "hardened" cookie on auth (signUp/login) and the hash of this fingerprint on the claim "X-User-Hashed-Fingerprint".
```ts
// Add the fingerprint in a hardened cookie to prevent Token Sidejacking
        // https://cheatsheetseries.owasp.org/cheatsheets/JSON_Web_Token_for_Java_Cheat_Sheet.html#token-sidejacking
        res.setHeader(
            "Set-Cookie",
            serialize(FINGERPRINT_COOKIE_NAME, fingerprint, {
                path: "/",
                maxAge: MAX_AGE,
                httpOnly: true,
                secure: process.env.NODE_ENV === "production",
            })
        )

        // Create JWT claims payload, including a hashed version of the 
        // fingerprint for the client to send back when asking for a token refresh
        const payload = {
            "https://hasura.io/jwt/claims": {
                "X-Hasura-Allowed-Roles": ["user"],
                "X-Hasura-Default-Role": "user",
                "X-Hasura-User-Id": String(user.id),
                // Token Sidejacking:
                // "A SHA256 hash of the random string will be stored in the token
                // (instead of the raw value) in order to prevent any XSS issues allowing the attacker
                // to read the random string value and setting the expected cookie."
                "X-User-Hashed-Fingerprint": sha256(fingerprint),
            },
        }

        // Sign the token, using HMAC SHA256 and give it a short duration
        // (Silent refresh will be responsible for repeated re-authing)
        const token = jwt.sign(payload, JWT_SECRET, {
            algorithm: "HS256",
            expiresIn: "5m",
        })
```
2. On request of a new refreshToken, the fingerprintHash from the claim is sent along with the current refreshToken:
```ts
const jwt = decodeJWT(getJwtToken())
const refreshToken = getRefreshToken()
const fingerprintHash = jwt?.["https://hasura.io/jwt/claims"]?.["X-User-Hashed-Fingerprint"]
```
 
3. The fingerprintHash from the claim and the hashed "hardened" fingerprint cookie are checked against before a new refreshToken is returned:
```ts
app.post("/api/actions/refresh-jwt", async (req, res) => {
    try {
        const { refreshToken, fingerprintHash } = req.body.input

        // First, check for the existence of the fingerprint HttpOnly cookie
        const fingerprintCookie = req.cookies[FINGERPRINT_COOKIE_NAME]
        if (!fingerprintCookie) return res.status(400).json({ message: "Unable to refresh JWT token" })

        // Compute a SHA256 hash of the received fingerprint in cookie in order to compare
        // it to the fingerprint hash stored in the token
        const fingerprintCookieHash = sha256(fingerprintCookie)

        // If the fingerprints don't match, the refresh is not authorized and is potentially malicious
        if (fingerprintHash != fingerprintCookieHash) {
            return res.status(400).json({ message: "Unable to refresh JWT token" })
        }

        // If the fingerprints do match, then continue to look the user up via their refresh token
        const user = await findUser({ refresh_token: { _eq: refreshToken } })
        if (!user) return res.status(400).json({ message: "User not found" })

        // Update user refresh token and refresh token expiration
        await updateUserRefreshToken({
            id: user.id,
            refresh_token: uuidv4(),
            refresh_token_expires_at: new Date(Date.now() + 1000 * 60 * 60 * 1).toISOString(),
        })


        // Sign and return new JWT
        const payload = {
            "https://hasura.io/jwt/claims": {
                "X-Hasura-Allowed-Roles": ["user"],
                "X-Hasura-Default-Role": "user",
                "X-Hasura-User-Id": String(user.id),
                "X-User-Hashed-Fingerprint": sha256(fingerprint),
            },
        }

        const token = jwt.sign(payload, JWT_SECRET, {
            algorithm: "HS256",
            expiresIn: "5m",
        })

        return res.json({ jwt: token })
    } catch (error) {
        console.log("/api/actions/refresh-jwt error", error)
        return res.status(400).json({ message: "Error issuing jwt token refresh" })
    }
})
``` 
An attacker could steal anything stored on sessionStorage or localStorage with JavaScript, but he couldn't steal the cookie that go along, which make the JWT auth more secure.

**Comment by osseonews** [comment]
"The Nhost SDK does not store the JWT but the refresh token in localStorage. We don't store JWTs anywhere else than in memory."

Isn't the benefit of JWT that it is a fast method of verification because you can store the JWT locally? If the user has to constantly get a new JWT on each page load by sending the refresh token to the server and waiting to get a JWT back from the server, won't this greatly effect the performance of the app?

Plus, if you are not storing the JWT in the cookie, how you are supposed to do verification on a server side endpoint? For example, NextJs has the pages/api folder. We send requests there. Generally, the JWT is sent as cookie to the api endpoint, and we can verify it in the backend, and then send a request to Hasura backend using the JWT. But, if no JWT is stored in the cookie, how are you supposed to do verification of API routes in NextJs?

Thanks.

**Comment by elitan** [comment]
We used cookies a long time ago with Hasura Backend Plus and we had so many issues with it. Mainly because browsers are starting to block third-party cookies. 

In general, if your app is vulnerable to XSS you should fix that first. Luckily, modern frameworks like React and Vue are protected against XSS by default.

> Isn't the benefit of JWT that it is a fast method of verification because you can store the JWT locally? If the user has to constantly get a new JWT on each page load by sending the refresh token to the server and waiting to get a JWT back from the server, won't this greatly affect the performance of the app?

The user doesn't have to generate a new JWT per page load. A JWT token is stored, in memory, for 15 minutes before being refreshed automatically in the background by our [JavaScript SDK](https://docs.nhost.io/reference/javascript#tokens-and-permissions).

> Plus, if you are not storing the JWT in the cookie, how you are supposed to do verification on a server-side endpoint?

Our JavaScript SDK makes sure to send the JWT token as an `Authorization` header.

If you're using our Serverless Functions you can also use the JWT token to verify the user making the request, check our guide: [Get the authenticated user in a Serverless Function](https://github.com/nhost/nhost/discussions/278).

**Comment by osseonews** [comment]
@elitan Thanks for the detailed response. I still have quite a few questions as I'm still a bit confused by the implementation here. We are using NextJs, so the questions below are based on that. Sorry for so many questions, but I didn't want to open an issue for each one, so hopefully we can address all of them here.

1.  "A JWT token is stored, in memory, for 15 minutes before being refreshed automatically in the background." Does this automatic refresh only happen if the entire app is wrapped with the NhostNextProvider? Otherwise, how does this actually occur? 

2.  "The user doesn't have to generate a new JWT per page load." Please see issue 622 (https://github.com/nhost/nhost/issues/622). We are using nhost cloud and nextjs and still have this same issue. We are running the latest versions of everything, but sometimes out of nowhere the refresh token requests start firing like crazy, exactly as documented in that issue.  I can't quite figure out what triggers this, so I wouldn't know how to reliably recreate the issue to debug it, but it happens alot.  

3. Assuming we don't use the nhost client, is there a function to generate a new access token from the refresh token without using the nhost client? I am asking b/c I am leaning towards using NextAuth in our app, as the implementation is a bit easier to understand, but in that case, we would need create a callback to generate a new access token after 15 minutes. I couldn't find anything in your docs for a function that will get a new access token with a refresh token.

4. Isn't there a security risk of exposing the backend nhost URL in every request when using the nhost client? Why should we be letting users know what our backend URL is to access our hasura instance? To get around this issue on our current hasura implementations (using hasura cloud), we simple "proxy" the request to hasura via a server, so the client only sees the API endpoint we proxy thru and not the actual hasura endpoint. Isn't this a more secure way to do it? But, the nhost client doesn't work this way, obviously, and it just calls the hasura endpoint directly. Which brings to the next question...

5. Serverless functions: As we are using NextJs we don't have a need for using Nhost serverless functions since NextJS can run serverless via the pages/api folder. In that case, would the guide still work. For example, would we just call our NextJS backend with await nhost.functions.call('/api/some-function', { name: 'my name'}); and the client will send in the token as an authorization header? It's not clear if the functions.call is only usable by nhost serverless functions or by any serverless function platform. My other idea was just to get the access token ourselves and pass it via a header using fetch. what's the benefit anyway of using nhost.functions.call instead of a just fetch to call a serverless function?

6. Serverless: In general, what I don't understand is how to use nhost in a Node environment anyway. It seems as if all guides are just use the nhost client, which is a client side library. Let's say we need to interact with nhost on the server side. We don't want to use hasura admin secret to interact with database, so how do we execute a call? Is the general idea to just get the JWT from the req auth header, and then attach it again in another request to the hasura backend as bearer token? Also, couldn't we create our own JWT to interact with the backend as the JWT secret is provided by Nhost?

7. Third party Cookies:  I'm not really sure I understand why there is a need to 3rd party cookies here. See how nextauth works.  The auth code basically runs on the server where the app is, so there is no 3rd party cookies; the cookie is set by the same server where the app is running. I presume you didn't go this route b/c you wanted something that could run even without a server and only in the client. But, now you offer serverless as part of nhost anyway, so I would think implementation with cookies would be much, much easier. I would think storing the access token in memory vs cookies (set by the app itself, and not a third party) would be prone to more issues than cookies and makes everything a bit more complicated.

**Comment by elitan** [comment]
No worries. I'll try my best to answer.

> "A JWT token is stored, in memory, for 15 minutes before being refreshed automatically in the background." Does this automatic refresh only happen if the entire app is wrapped with the NhostNextProvider? Otherwise, how does this actually occur?

This happens internally when you initialize the `NhostClient`. No provider is required.

```js
const nhost = new NhostClient({
  /* ... */
});
```

> "The user doesn't have to generate a new JWT per page load." Please see issue 622 (Next.js SDK fails to get refresh token  #622). We are using nhost cloud and nextjs and still have this same issue. We are running the latest versions of everything, but sometimes out of nowhere the refresh token requests start firing like crazy, exactly as documented in that issue. I can't quite figure out what triggers this, so I wouldn't know how to reliably recreate the issue to debug it, but it happens a lot.

We're just releasing a new version of Auth to Nhost Cloud. Let's see if that solves the issue. 

If token requests start firing like crazy that could indicate that you're initializing the `NhostClient` multiple times, as each `NhostClient` will refresh its internal tokens.

> Assuming we don't use the nhost client, is there a function to generate a new access token from the refresh token without using the nhost client? I am asking b/c I am leaning towards using NextAuth in our app, as the implementation is a bit easier to understand, but in that case, we would need create a callback to generate a new access token after 15 minutes. I couldn't find anything in your docs for a function that will get a new access token with a refresh token.

Yes, there is. It's all HTTP requests. Check the swagger spec for Auth here: https://docs.nhost.io/reference/hasura-auth/api-reference

> Isn't there a security risk of exposing the backend nhost URL in every request when using the nhost client? Why should we be letting users know what our backend URL is to access our hasura instance? To get around this issue on our current hasura implementations (using hasura cloud), we simple "proxy" the request to hasura via a server, so the client only sees the API endpoint we proxy thru and not the actual hasura endpoint. Isn't this a more secure way to do it? But, the nhost client doesn't work this way, obviously, and it just calls the hasura endpoint directly. Which brings to the next question...

No, there is no security risk exposing the backend Nhost URL. We use the same setup for our own Nhost Dashboard, as we give to customers. The Hasura instance is secured using the admin secret.

> Serverless functions: As we are using NextJs we don't have a need for using Nhost serverless functions since NextJS can run serverless via the pages/api folder. In that case, would the guide still work. For example, would we just call our NextJS backend with await nhost.functions.call('/api/some-function', { name: 'my name'}); and the client will send in the token as an authorization header? It's not clear if the functions.call is only usable by nhost serverless functions or by any serverless function platform. My other idea was just to get the access token ourselves and pass it via a header using fetch. what's the benefit anyway of using nhost.functions.call instead of a just fetch to call a serverless function?

The guide would work with some tweaking:

1. You need to manually set the JWT secret as an environment variable so it's available for your functions.
2. You can not use `nhost.functions.call()` as it will only call Serverless Functions hosted on Nhost.
3. When making a request to your functions in Next.js you need to manually send the access token as a header. You can access the access token using `nhost.auth.getAccessToken()`.

> Serverless: In general, what I don't understand is how to use nhost in a Node environment anyway. It seems as if all guides are just use the nhost client, which is a client side library. Let's say we need to interact with nhost on the server side. We don't want to use hasura admin secret to interact with database, so how do we execute a call? 

You can use the Nhost Client in a Node environment or in our Serverless Functions. Both being server-side. You can also make "raw" HTTP requests to Hasura, Auth, Storage, Serverless Functions.

> Is the general idea to just get the JWT from the req auth header, and then attach it again in another request to the hasura backend as bearer token?

Yes, passing the access (JWT) token as a header works. We use it sometimes. But it really depends on your use case.

>  Also, couldn't we create our own JWT to interact with the backend as the JWT secret is provided by Nhost?

Yes, you can. You can create your own JWT tokens using the `NHOST_JWT_SECRET` [system environment variable](https://docs.nhost.io/platform/environment-variables#system-environment-variables).

> Third party Cookies: I'm not really sure I understand why there is a need to 3rd party cookies here. See how nextauth works. The auth code basically runs on the server where the app is, so there is no 3rd party cookies; the cookie is set by the same server where the app is running. I presume you didn't go this route b/c you wanted something that could run even without a server and only in the client. But, now you offer serverless as part of nhost anyway, so I would think implementation with cookies would be much, much easier. I would think storing the access token in memory vs cookies (set by the app itself, and not a third party) would be prone to more issues than cookies and makes everything a bit more complicated.

The reason we need to account for third-party cookies is that the frontend runs on a different domain than the Nhost backend. To send requests with cookies from the frontend (Domain A) to the Nhost backend (Domain B) requires third-party cookies to be enabled.

From my understanding, NextAuth does not have this problem because they use the `pages/api/...` route in Next.js which lives on the same domain. Meaning, that both the frontend and the API folder are served from the same domain.

**Comment by plmercereau** [comment]
interesting, moving this issue to a discussion

**Comment by osseonews** [comment]
Thanks so much for the detailed responses. I am curious what you think of the strategy below for using NextJs functions. Per above, we can get the JWT token client side and pass as a header to the server, but in theory isn't it possible that by the time the serverless function runs that JWT will be expired? 

To prevent that type of situation, what I was thinking is that instead of passing the token, we just get the user's ID client side, and then pass that in the header instead. Then on the server side, we do a quick database check to make sure that user Id exists. If it does, we use the JWT secret to create a short-lived token (2 minutes let's say) to run some database operation for the user.

I know this seems like overkill, but won't this solve the problem of a potential "timeout" of the client side JWT token. Also, it seems quite secure, as I think it would be impossible for anyone to actually guess a user ID.

**Reply by elitan** [reply]
It's a trade-off. Do you want to either:

1. Risk that the JWT expires, or...
2. Risk that someone guess (or find) a user id?

I would rather go with option 1 and risk that the JWT expires. The risk of option 2 is too big in my opinion.

Also, the NhostClient updates the JWT token every 14 minutes (1 minute before it expires) so unless you run >1 min long operations in your function I would not be worried.

**Comment by ownadi** [comment]
I like the idea of passing the JWT to the nextjs API endpoint. How do I set a predictable JWT secret on development, though? `NHOST_JWT_SECRET={"type":"HS256", "key": "foo"}` in `.env.development` doesn't work. I need to access the predictable secret in the nextjs app.

**Reply by elitan** [reply]
I think that should work. Maybe there is some more information about it in the Next.js docs: https://nextjs.org/docs/basic-features/environment-variables

**Reply by ownadi** [reply]
@elitan, ty - I know how to set a variable in the Next.js project. The problem is that setting `NHOST_JWT_SECRET` has no effect in dev environment (`nhost dev` in CLI).

**Reply by ownadi** [reply]
Also, it looks like the verify endpoint from hasura-auth 0.8.0 could be used for this purpose w/o sharing the secret. It would be slower, tho.

**Reply by ownadi** [reply]
I've found in [the source code](https://github.com/nhost/cli/blob/2588ddafe8c685db59df5011a511bb7f16bbbf7e/util/vars.go#L5) that it's a hardcoded value. I can use it for developing the Next.js app and it works.

That being said, it would be nice to have an ability to set this `JWT_KEY` from an env variable, since the value is undocumented, non foobar-like, and may tend to change in the future.

**Reply by elitan** [reply]
I think during local development you can use it as a hard-coded value for now.

In production, you can find the jwt secret in the Nhost dashboard under Environment Variables -> NHOST_JWT_SECRET -> Reveal.
